### PR TITLE
Improve performance of Span.Fill

### PR DESCRIPTION
**Pull request by jkotas**
- Avoid taking address of the fill value
- Avoid unnecessary 64->32->64 bit integer truncate by doing all arithmetic as uint

### CODE REVIEW

**Comment by jkotas** [comment]
E.g. Span<char>.Fill is ~25% faster.

Code of the unrolled loop before:

```
movsxd  r9,ecx
mov     r10d,dword ptr [rsp+10h]
mov     word ptr [rdx+r9*2],r10w
lea     r9d,[rcx+1]
movsxd  r9,r9d
mov     r10d,dword ptr [rsp+10h]
mov     word ptr [rdx+r9*2],r10w
lea     r9d,[rcx+2]
movsxd  r9,r9d
mov     r10d,dword ptr [rsp+10h]
mov     word ptr [rdx+r9*2],r10w
lea     r9d,[rcx+3]
movsxd  r9,r9d
mov     r10d,dword ptr [rsp+10h]
mov     word ptr [rdx+r9*2],r10w
lea     r9d,[rcx+4]
movsxd  r9,r9d
mov     r10d,dword ptr [rsp+10h]
mov     word ptr [rdx+r9*2],r10w
lea     r9d,[rcx+5]
movsxd  r9,r9d
mov     r10d,dword ptr [rsp+10h]
mov     word ptr [rdx+r9*2],r10w
lea     r9d,[rcx+6]
movsxd  r9,r9d
mov     r10d,dword ptr [rsp+10h]
mov     word ptr [rdx+r9*2],r10w
lea     r9d,[rcx+7]
movsxd  r9,r9d
mov     r10d,dword ptr [rsp+10h]
mov     word ptr [rdx+r9*2],r10w
add     ecx,8
cmp     ecx,r8d
jl      again
```

After:

```
mov     word ptr [rcx+r8*2],dx
mov     word ptr [rcx+r8*2+2],dx
mov     word ptr [rcx+r8*2+4],dx
mov     word ptr [rcx+r8*2+6],dx
mov     word ptr [rcx+r8*2+8],dx
mov     word ptr [rcx+r8*2+0Ah],dx
mov     word ptr [rcx+r8*2+0Ch],dx
mov     word ptr [rcx+r8*2+0Eh],dx
add     r8,8
cmp     r8,r9
jb      again
```

**Comment by davidfowl** [comment]
How much faster is it?

**Comment by redknightlois** [comment]
Besides the improvement itself, which is great; the codegen trick is way better!! ❤️

**Comment by ahsonkhan** [comment]
That's awesome. LGTM :)

File: src/mscorlib/src/System/Span.cs [hunk 1]
```diff
@@ -224,47 +224,50 @@ public void Clear()
...
             if (Unsafe.SizeOf<T>() == 1)
             {
-                byte fill = Unsafe.As<T, byte>(ref value);
-                ref byte r = ref Unsafe.As<T, byte>(ref _pointer.Value);
-                Unsafe.InitBlockUnaligned(ref r, fill, (uint)length);
+                uint length = (uint)_length;
+                if (length == 0)
+                    return;
+
+                T tmp = value; // Avoid taking address of the "value" argument. It would regress performance of the loop below.
```
Comment (ahsonkhan) [inline src/mscorlib/src/System/Span.cs]: Can you explain this further?

Comment (jkotas) [inline src/mscorlib/src/System/Span.cs]: If you take an address of a local variable or argument, it needs to get spilled to the stack and every access to it gets slower.

**Pull request merged**